### ISSUE TITLE
Netty javax.net.ssl.SSLException: SSLEngine closed already error

### ISSUE THREAD

**Issue by maybhi** [netty/netty#6228]
### Expected behavior
No Error
### Actual behavior
I am getting error when i write data at client side using 

ChannelFuture future=channel.writeAndFlush(uMessage);
				future.addListener(new GenericFutureListener<Future<? super Void>>() {

					@Override
					public void operationComplete(Future<? super Void> future) throws Exception {
						// TODO Auto-generated method stub
						if(future.isSuccess()){
							System.out.println("Success");
						}
						else{
							System.out.println("Failure");
							future.cause().printStackTrace();
						}
					}
				});

I get below error from future.cause().printStackTrace();

javax.net.ssl.SSLException: SSLEngine closed already
	at io.netty.handler.ssl.SslHandler.wrap(...)(Unknown Source)

Here this is my client code,
public SSLEngine getSslEngine() throws Exception {

		TrustManagerFactory tmFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
		String keystorePass = "123456";

		KeyStore ks = KeyStore.getInstance("JKS");
		tmFactory.init(ks);

		ks.load(new FileInputStream("resources/newssl/keystore.jks"), keystorePass.toCharArray());

		// Set up key manager factory to use our key store
		KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
		kmf.init(ks, keystorePass.toCharArray());

		KeyManager[] km = kmf.getKeyManagers();
		TrustManager[] tm = tmFactory.getTrustManagers();

		SSLContext sslContext = SSLContext.getInstance("TLS");
		sslContext.init(km, tm, null);
		sslContext.getServerSessionContext().setSessionCacheSize(0);
		sslContext.getServerSessionContext().setSessionTimeout(0);
		SSLEngine sslEngine = sslContext.getDefault().createSSLEngine(serverIp, serverPort);

		sslEngine.setUseClientMode(true);
		sslEngine.setEnabledProtocols(sslEngine.getSupportedProtocols());
		sslEngine.setEnabledCipherSuites(sslEngine.getSupportedCipherSuites());
		return sslEngine;
	}

Here its my server code,

TrustManagerFactory tmFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
		KeyStore ks = KeyStore.getInstance("JKS");
		tmFactory.init(ks);
		String keystorePass = "123456";

		ks.load(new FileInputStream("src/main/resources/newssl/keystore.jks"), keystorePass.toCharArray());

		// Set up key manager factory to use our key store
		KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
		kmf.init(ks, keystorePass.toCharArray());

		KeyManager[] km = kmf.getKeyManagers();
		TrustManager[] tm = tmFactory.getTrustManagers();

		SSLContext sslContext = SSLContext.getInstance("TLS");
		sslContext.init(km, tm, null);
		sslContext.getServerSessionContext().setSessionCacheSize(0);
		sslContext.getServerSessionContext().setSessionTimeout(0);
		SSLEngine sslEngine = sslContext.getDefault().createSSLEngine("127.0.0.1", 11235);
		sslEngine.setUseClientMode(false);
		sslEngine.setEnabledProtocols(sslEngine.getSupportedProtocols());
		sslEngine.setEnabledCipherSuites(sslEngine.getSupportedCipherSuites());
		sslEngine.setEnableSessionCreation(true);
		return sslEngine;
 
### Netty version
4.1.6.Final
### JVM version (e.g. `java -version`)
JavaSE-1.7
### OS version (e.g. `uname -a`)
Windows Server 2012 R2 Standard

**Comment by johnou** [comment]
@maybhi I think this was fixed with netty-4.1.7.Final

**Comment by maybhi** [comment]
@johnou I checked it. Its not working in netty 4.1.7.Final. I am gettting same error.

**Comment by normanmaurer** [comment]
@maybhi I suspect that your handshake failed and so the Channel is closed.

did you check `sslHandler.handshakeFuture()` ?

**Comment by maybhi** [comment]
@normanmaurer I checked with below code
                 SSLEngine sslEngine = client.getSslEngine();
		socketChannel.pipeline().addFirst("handler", new SslHandler(sslEngine, false));
               Future f1=sslHandler.handshakeFuture();
		f1.addListener(new GenericFutureListener<Future<?>>() {
			
			@Override
			public void operationComplete(Future<?> future) throws Exception {
			  if(future.isSuccess()){
				  System.out.println("Success");
			  }
			  else{
				  System.out.println("Failure---");
			  }
				
			}
		} );

and i am getting Failure , so what could be the reason, how to resolve it?

**Comment by normanmaurer** [comment]
@maybhi what is the exception ?

```java
    future.cause().printStracktrace():
```

**Comment by maybhi** [comment]
now handshake is ok. It gives me Success output but when i write to the channel i got the same error the code is as per below

if (channel.isActive()) {
				System.out.println("before write "+input);
				ChannelFuture future=channel.writeAndFlush(uMessage);
				future.addListener(new GenericFutureListener<Future<? super Void>>() {

					@Override
					public void operationComplete(Future<? super Void> future) throws Exception {
						// TODO Auto-generated method stub
						if(future.isSuccess()){
							System.out.println("Success");
						}
						else{
							System.out.println("Failure while writing");
							future.cause().printStackTrace();
							
						}
					}
				});
			} else {
				System.out.println("channel closed");
			}

i am getting Failure while writing output

**Comment by normanmaurer** [comment]
@maybhi please share a complete reproducer that I can run locally

**Comment by maybhi** [comment]
@normanmaurer I got the point and solved it. As i had not written ctx.flush(); in channelReadComplete method of server handler. When i did, it works. and if i remove that line again i get the same error. so need to write ctx.flush in channelReadComplete method.

**Comment by maybhi** [comment]
@Scottmitch Please tell me why shoud we write ctx.flush(); in channelReadComplete method of server handler to avoid the error what we get?

**Comment by johnou** [comment]
@maybhi from the snippets you have provided it's hard to get context, would you please provide a full reproducer?

**Comment by unoexperto** [comment]
I'm on `4.1.9.Final` and still getting this error. I use Netty via http://mvnrepository.com/artifact/org.asynchttpclient/async-http-client/2.1.0-alpha12

**Comment by johnou** [comment]
@unoexperto please create a new issue with all information to your case, reproducer would be fantastic.

**Comment by unoexperto** [comment]
@johnou It's unlikely I'm gonna be able to do that (and I saw at least 3 other tickets about this error). It happens only after 8-10 hours of constant work of the code. If you tell me how I can collect necessary information when exception occurs I'll do that.

**Comment by Scottmitch** [comment]
@maybhi - As @johnou said it is difficult to diagnose the issue without a reproducer. Please reproducer that we can run locally so we can see what is actually going on.

**Comment by ostronom** [comment]
This issue is closed, but I'm running into it with 4.1.21.Final. 

Connection fails with this exception (with loglevel 7 for some reason), and after that -- no exception are thrown (connection looks like alive), but no data can be sent through it -- data is silently swallowed.
